src/main/java/com/android/volley/toolbox/BasicAsyncNetwork.java

```diff
@@ -0,0 +1,432 @@
+                                        callback,
+                                        e,
+                                        requestStartMs,
+                                        /* httpResponse= */ null,
+                                        /* responseContents= */ null);
+                            }
+                        }
+                    });
+        }
+    }
+
+    /**
+     * This method sets the non blocking executor to be used by the stack for non-blocking tasks. If
+     * you are not using an {@link com.android.volley.toolbox.AsyncHttpStack}, this should do
+     * nothing. This method must be called before performing any requests if you are using an
+     * AsyncHttpStack.
+     */
+    @Override
+    public void setNonBlockingExecutorForStack(ExecutorService executor) {
+        if (mBaseHttpStack instanceof AsyncHttpStack) {
+            AsyncHttpStack stack = (AsyncHttpStack) mBaseHttpStack;
+            stack.setNonBlockingExecutor(executor);
+        } else {
+            VolleyLog.d("Cannot set non-blocking executor for non-async stack");
+        }
+    }
+
+    /**
+     * This method sets the blocking executor to be used by the network and stack for potentially
+     * blocking tasks. This method must be called before performing any requests. Only set the
+     * executor for the stack if it is an instance of {@link
+     * com.android.volley.toolbox.AsyncHttpStack}
+     */
+    @Override
+    public void setBlockingExecutor(ExecutorService executor) {
+        mBlockingExecutor = executor;
+        if (mBaseHttpStack instanceof AsyncHttpStack) {
+            AsyncHttpStack stack = (AsyncHttpStack) mBaseHttpStack;
+            stack.setBlockingExecutor(executor);
+        } else {
+            VolleyLog.d("Cannot set blocking executor for non-async stack");
+        }
+    }
+
+    /**
+     * Attempts to prepare the request for a retry. If there are no more attempts remaining in the
+     * request's retry policy, a timeout exception is thrown.
+     *
+     * @param request The request to use.
+     */
+    private void attemptRetryOnException(
+            final String logPrefix,
+            final Request<?> request,
+            final OnRequestComplete callback,
+            final VolleyError exception) {
+        final RetryPolicy retryPolicy = request.getRetryPolicy();
+        final int oldTimeout = request.getTimeoutMs();
+
+        try {
+            retryPolicy.retry(exception);
+        } catch (VolleyError e) {
+            request.addMarker(
+                    String.format("%s-timeout-giveup [timeout=%s]", logPrefix, oldTimeout));
+            callback.onError(e);
+            return;
+        }
+        request.addMarker(String.format("%s-retry [timeout=%s]", logPrefix, oldTimeout));
+        // If we are using an async stack, then perform retry after a timeout.
+        if (mBaseHttpStack instanceof AsyncHttpStack) {
+            mHandler.postDelayed(
```

Comment:
We shouldn't need a Handler for this; we should be able to use the Executor. But we might need to change it to a ScheduledExecutorService instead if we decide to keep this behavior.

Reply:
If what we're passing in is a ThreadPoolExecutor, how could we change it to a ScheduledExecutorService?

Reply:
We'd need to use something different than a ThreadPoolExecutor in that case. ScheduledThreadPoolExecutor should work.

(That said, this may not be a problem for now if we drop the timing constraint, although we may want to keep the API as requiring a ScheduledExecutorService so that we'll have that option in the future).